### PR TITLE
Added LruCache in envoy

### LINKED ISSUE #16108: move simple LRU cache code into envoy
In this [PR](https://github.com/envoyproxy/envoy/pull/14341), jwt_authn is adding cache feature JWT token.  

It is using a simple_LRU_cache map from [here](https://github.com/google/jwt_verify_lib/tree/master/simple_lru_cache)

This map supports O(1) lookup with LRU auto purge.

This simple_LRU_cache code can be used by other Envoy modules, hence propose to move this code into Envoy.

### PR DESCRIPTION
**Pull request by mk46**
Signed-off-by: Manish Kumar <manish.kumar1@india.nec.com>

Commit Message: Move LruCache from [jwt_verify_lib/simple_lru_cache](https://github.com/google/jwt_verify_lib/tree/master/simple_lru_cache)
Additional Description: This PR is follow up to [comment:14341](https://github.com/envoyproxy/envoy/pull/14341/files#r617951885)
Risk Level:
Testing: unit and build
Docs Changes: n/a
Release Notes: n/a
Fixes #16108

### CODE REVIEW

**Comment by mk46** [comment]
CC @wrowe @htuch @qiwzhang

**Comment by tonya11en** [comment]
Also, why not just add https://github.com/google/jwt_verify_lib/tree/master/simple_lru_cache as an external dependency rather than copy the files?

**Comment by mk46** [comment]
> Also, why not just add https://github.com/google/jwt_verify_lib/tree/master/simple_lru_cache as an external dependency rather than copy the files?

As of now, this is already used as an external dependency in [Jwt Cache](https://github.com/envoyproxy/envoy/pull/14341). The purpose of adding to envoy is to consumed by other modules of envoy.

This code has a copyright of google. Please see https://github.com/google/jwt_verify_lib/blob/e5d6cf7067495b0868787e1fd1e75cef3242a840/simple_lru_cache/simple_lru_cache.h#L1

**Comment by mk46** [comment]
@htuch Any thought?

**Comment by htuch** [comment]
I think the idea was that this might be useful beyond the JWT verifier. It seems there are a few other places that LRU might make sense (basically, whenever we're caching stuff, e.g. DNS, gRPC clients, etc.). Maybe do a quick survey and list all places LRU cache might make sense based on existing uses?

**Comment by mk46** [comment]
Initially, I had tried to keep this code in envoy but as per discussion in https://github.com/envoyproxy/envoy/pull/14341#issuecomment-743986331, moved Lru cache code in jwt_verify_lib.

File: source/common/lru_cache/lru_cache_inl.h [hunk 1]
```diff
@@ -0,0 +1,1046 @@
...
+#include <utility>
+
+#include "source/common/lru_cache/lru_cache.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace LruCache {
+
+#undef GOOGLE_DISALLOW_EVIL_CONSTRUCTORS
```
Comment (yanavlasov) [inline source/common/lru_cache/lru_cache_inl.h:49]: Please use the `= delete ;` syntax to make class noncopyable without the macro.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

**Comment by yanavlasov** [comment]
Reason for the format check failure: https://dev.azure.com/cncf/envoy/_build/results?buildId=83500&view=logs&j=c5dd2866-6ab3-5f3c-3a44-4cef0ec909b5&t=a9eb66d6-8944-5769-b3f7-476949dadcb8&l=72

**Review by qiwzhang — APPROVED** [review]

File: source/common/lru_cache/BUILD [hunk 2]
```diff
@@ -0,0 +1,17 @@
+load(
```
Comment (lizan) [inline source/common/lru_cache/BUILD:1]: IMO we can just put this in `source/common/common`

File: source/common/lru_cache/lru_cache_inl.h [hunk 3]
```diff
@@ -0,0 +1,1046 @@
...
+#define GOOGLE_DISALLOW_EVIL_CONSTRUCTORS(TypeName)                                                \
+  TypeName(const TypeName&);                                                                       \
+  void operator=(const TypeName&)
+
+// Define number of microseconds for a second.
+const int64_t kSecToUsec = 1000000;
```
Comment (lizan) [inline source/common/lru_cache/lru_cache_inl.h:55]: constexpr

**Review by lizan — CHANGES_REQUESTED** [review]
please adjust coding style as well.

**Comment by mk46** [comment]
I am closing as of now, I have been busy with some other task.

**Pull request closed without merging**